client/src/lib/dependency-injection/ApolloGraphqlManager.js

```diff
@@ -0,0 +1,312 @@
+const protectedConfig = ['templateName', 'fields', 'params', 'fragments'];
+const deferredApolloConfig = ['options', 'props', 'variables', 'skip', 'update'];
+
+/**
+ * An API for updating the query/mutation parts of a given template
+ */
+class ApolloGraphqlManager {
+  /**
+   * @schema for the config structure
+    {
+      apolloConfig: {
+        // additive only, will ignore removals
+        options: (<originalParams>) => (<currentValue>) => <object|newValue>,
+
+        // additive only, will ignore removals
+        props: (<originalData>) => (<currentValue>) => <object|newValue>,
+
+        // additive only, will ignore removals
+        variables: (<props>) => (<currentValue) => <object|newValue>,
+
+        // last transform returning a boolean will take precedence
+        skip: (<props>) => (<currentValue>) => <boolean|newValue>,
+
+        // additive only, will ignore removals??
+        context: (currentValue) => <object|newValue>,
+
+        // additive only, will ignore removals
+        optimisticResponse: (<currentValue>) => <object|newValue>,
+
+        // additive only, will ignore removals
+        refetchQueries: (<currentValue>) => <array|newValue>,
+
+        // last transform returning a number will take precedence
+        pollInterval: (<currentValue>) => <integer|newValue>,
+
+        // last transform returning an object or null will take precedence
+        fetchPolicy: (<currentValue>) => <object|newValue>,
+
+        // last transform returning a string will take precedence
+        name: (<currentName>) => <string|newValue>,
+
+        withRef: <anyTrue>,
+
+        notifyOnNetworkStatusChange: <anyTrue>,
+
+        // each transformation will be called successively no values returned between
+        // each transformation
+        update: (<DataProxy>, <Response>) => <noReturn>,
+      },
+      singularName: <string>,
+      pluralName: <string>,
+      fields: [<string>], - e.g. ['Title'] or nested fields ['Author', ['ID']]
+      params: {<name>: <type>}, - e.g. { limit: 'Int' }
+      pagination: <boolean>,
+      templateName: <string>,
+    }
+   *
+   * @param {object} config - default config options which we extrapolate as the "base" behaviour.
+   * @param {object} templates - a key-value map of templates
+   * @param {object} fragments - a key-value map of fragments
+   */
+  constructor(config, templates, fragments) {
+    const {
+      apolloConfig,
+      ...otherConfig
+    } = config;
+
+    this.config = otherConfig;
+    this.apolloConfigInitValues = apolloConfig;
+    this.apolloConfigTransforms = {};
+    this.templates = { ...templates } || {};
+    this.fragments = { ...fragments } || {};
+
+    this.reduceApolloConfig = this.reduceApolloConfig.bind(this);
+  }
+
+  setConfig(name, value) {
+    // prevent overriding functionally important configs
+    if (protectedConfig.includes(name)) {
+      throw new Error(`
+Tried to set protected config values: '${name}', which is discouraged.
```

Comment:
Weird newline here.

Reply:
I think all error messages should use something like this https://github.com/airbnb/deline to get rid of coding indents when it appears in the console

Those lengthy error messages plus indents is quite an eyesore now :D 

I can raise a pull request later on.